proto/asn1-pdu/x509_certificate_to_der.cc

```diff
@@ -43,6 +43,172 @@ DECLARE_ENCODE_FUNCTION(AlgorithmIdentifierSequence) {
   EncodeTagAndLength(kAsn1Sequence, der.size() - tag_len_pos, tag_len_pos, der);
 }
 
+DECLARE_ENCODE_FUNCTION(ExtendedKeyUsage) {
+  Encode(val.key_purpose_id(), der);
+  for (const auto& key_purpose_id : val.key_purpose_ids()) {
+    Encode(key_purpose_id, der);
+  }
+}
+
+DECLARE_ENCODE_FUNCTION(BasicConstraints) {
+  if (val.ca().val()) {
+    Encode(val.ca(), der);
+  }
+  if (val.has_path_len_constraint()) {
+    Encode(val.path_len_constraint(), der);
+  }
+}
+
+DECLARE_ENCODE_FUNCTION(KeyUsage) {
+  uint16_t key_usage = 0x00;
+  auto masks = {val.digital_signature() ? 0x01 : 0x00,
+                val.digital_signature() ? 0x02 : 0x00,
+                val.digital_signature() ? 0x04 : 0x00,
+                val.digital_signature() ? 0x08 : 0x00,
+                val.digital_signature() ? 0x10 : 0x00,
+                val.digital_signature() ? 0x20 : 0x00,
+                val.digital_signature() ? 0x40 : 0x00,
+                val.digital_signature() ? 0x80 : 0x00,
+                val.digital_signature() ? 0x100 : 0x00};
```

Comment:
BUG: This copy/pastes the same bool check, which probably isn't what you want?

You 'should' be able to get away with something like (and sorry, this is where coding in a comment box is terrible)

```suggestion
constexpr struct {
  using Fn = bool (KeyUsage::*)(void);
  Fn function;
  uint32_t mask;
} kValues[] = {
  {
    { &KeyUsage.digital_signature, 0x01 },
    { &KeyUsage.non_repudiation, 0x02 },
    // ...
   }
};
for (const auto& usage : kValues) {
  if ((val.*(usage.function))()) {
  // alt: if std::invoke(val, usage.function), but that's C++17 and best to avoid for now
    key_usage |= usage.mask;
  }
}
```

I think that's right. In the structure, we take pointers to the function to call and the mask to use. In the usage, we then dereference that function, call it, and if it's set, add the mask.